rados/write_op_exec.go

```diff
@@ -0,0 +1,82 @@
+//go:build ceph_preview
+// +build ceph_preview
+
+package rados
+
+// #cgo LDFLAGS: -lrados
+// #include <stdlib.h>
+// #include <rados/librados.h>
+//
+import "C"
+
+import (
+	"unsafe"
+)
+
+// writeOpExecStep - exec step in write operation.
+type writeOpExecStep struct {
+	prval *C.int
+}
+
+// newWriteOpExecStep - init new *writeOpExecStep.
+func newWriteOpExecStep() *writeOpExecStep {
+	es := &writeOpExecStep{
+		prval: (*C.int)(C.malloc(C.sizeof_int)),
+	}
+	*es.prval = 0
+
+	return es
+}
+
+// free - will be called to free any resources, especially C memory.
+func (es *writeOpExecStep) free() {
+	if es.prval != nil {
+		C.free(unsafe.Pointer(es.prval))
+		es.prval = nil
+	}
+}
+
+// update - update state operation.
+func (es *writeOpExecStep) update() error {
+	return getError(*es.prval)
+}
+
+// Exec - Execute an OSD class method on an object See rados_exec() for general description.
+//
+// Implements:
+//
+//	void rados_write_op_exec(rados_write_op_t write_op,
+//	                         const char *cls,
+//	                         const char *method,
+//	                         const char *in_buf,
+//	                         size_t in_len,
+//	                         int *prval)
+func (w *WriteOp) Exec(in []byte, clsName, method string) error {
+	cClsName := C.CString(clsName)
+	defer C.free(unsafe.Pointer(cClsName))
+
+	cMethod := C.CString(method)
+	defer C.free(unsafe.Pointer(cMethod))
+
+	var cInBuff *C.char
+	var cInBuffLen C.size_t
+
+	if len(in) > 0 {
+		ws := newWriteStep(in, 0, 0)
+		w.steps = append(w.steps, ws)
+		cInBuff = ws.cBuffer
+		cInBuffLen = ws.cDataLen
```

Comment:
Same as my earlier comment, I think that it would be cleaner to add cBuffer and cDataLen fields to the writeOpExecStep and create one exec step for this call than generate two exec steps when there is input.